apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ManagedTransaction.cs

```diff
@@ -0,0 +1,752 @@
+        /// </remarks>
+        public Timestamp ReadTimestamp => Interlocked.CompareExchange(ref _transaction, null, null)?.ReadTimestamp;
+
+        // internal for testing
+        internal MultiplexedSessionPrecommitToken PrecommitToken { get; set; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="multiplexSession"></param>
+        /// <param name="transactionId"></param>
+        /// <param name="transactionOptions"></param>
+        /// <param name="singleUseTransaction"></param>
+        /// <param name="readTimestamp"></param>
+        public ManagedTransaction(ManagedSession multiplexSession, ByteString transactionId, TransactionOptions transactionOptions, bool singleUseTransaction, Timestamp readTimestamp)
+        {
+            _multiplexSession = multiplexSession;
+
+            TransactionOptions = transactionOptions?.Clone() ?? new TransactionOptions();
+
+            GaxPreconditions.CheckArgument(
+                TransactionOptions.ModeCase == ModeOneofCase.ReadOnly || !singleUseTransaction,
+                nameof(singleUseTransaction),
+                "Single use transactions are only supported for read-only transactions.");
+            GaxPreconditions.CheckArgument(
+                transactionId is null || TransactionOptions.ModeCase != ModeOneofCase.None,
+                nameof(transactionOptions),
+                $"No transaction options were specified for the given transasaction ID {transactionId is null}, {transactionId?.ToBase64()}, {TransactionOptions.ModeCase}.");
+            GaxPreconditions.CheckArgument(
+                readTimestamp is null || transactionId is not null,
+                nameof(readTimestamp),
+                "A read timestamp can only be specified if a transaction ID is also being specified.");
+
+            SingleUseTransaction = singleUseTransaction;
+            if (transactionId is not null)
+            {
+                // We don't need to lock here, this is the constructor.
+                _transaction = new Transaction
+                {
+                    Id = transactionId,
+                    ReadTimestamp = readTimestamp,
+                };
+            }
+        }
+
+        internal ManagedTransaction WithTransaction(ByteString transactionId, TransactionOptions transactionOptions, bool singleUseTransaction, Timestamp readTimestamp = null)
+        {
+            return new ManagedTransaction(_multiplexSession, transactionId, transactionOptions, singleUseTransaction, readTimestamp);
+        }
+
+        private Mutation MaybeFetchMutationKey(Protobuf.Collections.RepeatedField<Mutation> mutations)
+        {
+            if(mutations.Count < 1)
+            {
+                return null;
+            }
+
+            // Fetch first mutation in list, we will return the first mutation by default if it meets all conditions
+            Mutation key = mutations.ElementAt(0);
```

Comment:
Something like this:

```suggestion
            mutations.Where(m => m.KeySet.Keys.Count > 0 || m.Delete?.KeySet.Kets.Count > 0)
```